reconcile/utils/external_resources.py

```diff
@@ -47,3 +52,122 @@ def managed_external_resources(namespace_info: Mapping[str, Any]) -> bool:
+                                    provider expects an `identifier` argument.
+        """
+        self._spec = spec
+        self._integration_tag = integration_tag
+        self._identifier_as_value = identifier_as_value
+
+    def resolve(self) -> dict:
+        """
+        Resolves the final set of values after aggregating and overriding values.
+        """
+        resource = self._spec.resource
+
+        keys_to_add = [
+            key for key in self._spec.resource.keys() if key not in self._IGNORE_KEYS
+        ]
+
+        defaults_path = resource.get("defaults", None)
+        overrides = resource.get("overrides", None)
+
+        # TODO: see Gerd's example in https://issues.redhat.com/browse/APPSRE-6003 and
+        # just handle the defaults values directly. This isn't a blocker for the initial
+        # PR because it needs to be updated on the integration side.
+        values = self._get_values(defaults_path) if defaults_path else {}
+        self._aggregate_values(values)
+        self._override_values(values, overrides)
+
+        if self._identifier_as_value:
+            values["identifier"] = self._spec.identifier
+
+        if self._integration_tag:
+            values["tags"] = self._spec.tags(self._integration_tag)
+
+        for key in keys_to_add:
+            val = resource.get(key, None)
+            # checking explicitly for not None
+            # to allow passing empty strings, False, etc
+            if val is not None:
+                values[key] = val
+
+        return values
+
+    def _get_values(self, path: str) -> dict:
+        raw_values = self._get_raw_values(path)
+        try:
+            values = anymarkup.parse(raw_values["content"], force_types=None)
+            values.pop("$schema", None)
+        except anymarkup.AnyMarkupError:
+            e_msg = "Could not parse data. Skipping resource: {}"
+            raise FetchResourceError(e_msg.format(path))
+        return values
+
+    @staticmethod
+    def _get_raw_values(path):
+        gqlapi = gql.get_api()
+        try:
+            raw_values = gqlapi.get_resource(path)
+        except gql.GqlGetResourceError as e:
+            raise FetchResourceError(str(e))
+        return raw_values
+
+    @staticmethod
+    def _aggregate_values(values):
+        split_char = "."
+        copy = values.copy()
+        for k, v in copy.items():
+            if split_char not in k:
+                continue
+            k_split = k.split(split_char)
+            primary_key = k_split[0]
+            secondary_key = k_split[1]
+            values.setdefault(primary_key, {})
+            values[primary_key][secondary_key] = v
+            values.pop(k, None)
+
+    @staticmethod
+    def _override_values(values, overrides):
+        if overrides is None:
+            return
+        data = json.loads(overrides)
+        for k, v in data.items():
+            values[k] = v
```

Comment:
We probably want to add type hints to these methods, even though they are a direct copy from TerrascriptAwsClient

Reply:
These were just accidentally missed. I've added the following to prevent missing types in new modules:

```
[mypy-reconcile.utils.external_resources.*]
disallow_untyped_defs = True
disallow_incomplete_defs = True
```